### PR TITLE
Offline Login and realm access Sample

### PR DESCRIPTION
**Pull request by dianaafanador3**

### CODE REVIEW

**Comment by dianaafanador3** [comment]
@dacharyc

File: OfflineLoginRealmAccess/README.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+## Scope
+
+The app addresses the following points:
+* Register/Login a Device Sync user.
+* Opening a Realm with and without and internet connection.
+* Add an initial flexible sync subscription, which executes when opening a 
+  realm for the very first time, even if offline.
+* Logging out a Device Sync user and resetting your SwiftUI's associated Realm when doing so.
+
+* Setting the default logger to `.all`, showing all the sync and realm access logs.
```
Comment (dacharyc) [inline OfflineLoginRealmAccess/README.md]: Not sure why there's an empty line here - if this isn't intentional, we should probably remove it.
```suggestion
* Logging out a Device Sync user and resetting your SwiftUI's associated Realm when doing so.
* Setting the default logger to `.all`, showing all the sync and realm access logs.
```

File: OfflineLoginRealmAccess/README.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+* iOS: 16.0
+
+## Getting Started
+
+### Set up an Atlas App Services App
+
+To sync Realm data you must first:
+
+1. [Create an App Services App](https://www.mongodb.com/docs/atlas/app-services/manage-apps/create/create-with-ui/)
+2. Enable [API Key Authentication](https://www.mongodb.com/docs/atlas/app-services/authentication/api-key/)
```
Comment (dacharyc) [inline OfflineLoginRealmAccess/README.md]: This app seems to use email/password auth, not API Key authentication.
```suggestion
2. Enable [Email/Password Authentication](https://www.mongodb.com/docs/atlas/app-services/authentication/email-password/#std-label-email-password-authentication)
```

File: OfflineLoginRealmAccess/OfflineLoginRealmAccess/Managers/RealmManager.swift [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+        // Setting the download configuration to `.once` will wait to sync data only the first time the realm
+        // is opened, otherwise this will open the realm immediately and will sync any data on the background
+        // if connected.
+        do {
+            // In case you want to sync your Device Sync data on every realm open, you can change this
+            // to `.always`, which in case of no internet connection will throw into opening the realm
+            // offline,
+            let realm = try await Realm(configuration: configuration, downloadBeforeOpen: .once)
+        } catch {
```
Comment (dacharyc) [inline OfflineLoginRealmAccess/OfflineLoginRealmAccess/Managers/RealmManager.swift]: When I tried to run the app, I was never getting past the ProgressView spinner. Moving the `self.realm = realm` into the `do` block fixed the issue. If you do take this suggestion, we should remove it from line 49 below, but I'm not sure this is the best fix.
```suggestion
            let realm = try await Realm(configuration: configuration, downloadBeforeOpen: .once)
            self.realm = realm
        } catch {
```

File: OfflineLoginRealmAccess/OfflineLoginRealmAccess/Managers/RealmManager.swift [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+import Foundation
+import RealmSwift
+
+class RealmManager: ObservableObject {
+    @Published var realm: Realm?
+
+    // Set the flexible sync configuration for the realm.
+    @MainActor
+    func openRealmWithUser(_ user: User) async throws {
```
Comment (dacharyc) [inline OfflineLoginRealmAccess/OfflineLoginRealmAccess/Managers/RealmManager.swift:27]: The structure of this func made me do a lot of extra thinking. The do/catch pattern where you perform a failable call in the catch block and then throw is not one I'm accustomed to seeing. A more "Swifty" approach might be to have a specific `catch` clause for the case where the user is offline, and in that case open where `downloadBeforeOpen: .never`. (I tried poking around but on a quick glance I'm not sure what that error is.) But I'm thinking something like:

```swift
do {
    let realm = try await Realm(configuration: configuration, downloadBeforeOpen: .once)
} catch SomeErrorIndicatingTheUserIsOffline {
   let realm = try await Realm(configuration: configuration, downloadBeforeOpen: .never)
} catch {
    print("There was an error opening the realm: \(error.localizedDescription)")
}
```

Just a thought.

File: OfflineLoginRealmAccess/OfflineLoginRealmAccess/Managers/RealmManager.swift [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        do {
+            // In case you want to sync your Device Sync data on every realm open, you can change this
+            // to `.always`, which in case of no internet connection will throw into opening the realm
+            // offline,
+            let realm = try await Realm(configuration: configuration, downloadBeforeOpen: .once)
+        } catch {
+            // In case there is an issue on the first sync, we will try to open the realm without
+            // syncing any data even on the first open.
+            // This will throw in case of the error been related to the configuration.
+            let realm = try await Realm(configuration: configuration, downloadBeforeOpen: .never)
```
Comment (dacharyc) [inline OfflineLoginRealmAccess/OfflineLoginRealmAccess/Managers/RealmManager.swift]: This particular example, of trying to `downloadBeforeOpen: .once` and then switching to `downloadBeforeOpen: .never` in a catch block where `.once` fails struck me as an odd example.

There _must_ be a logged-in user the first time the app opens, otherwise you never get past the login screen. So I'm not understanding a case where the app had an internet connection long enough to login but lost the internet connection before opening the realm. _Maybe_ this is a good example if there is a large initial download, and the connection might be lost during the initial sync? But if that is the case, would this example even work, or would that throw a different error?

With the current structure, I would assume the `let realm` in the `catch` block would _never_ get called.

File: OfflineLoginRealmAccess/OfflineLoginRealmAccess/ContentView.swift [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+        }
+    }
+
+    func addNewBook() {
+        let newBook = Book()
+        newBook.title = text
+        newBook.userId = user.id
+        newBook.isPublic = false
+        $books.append(newBook)
+    }
```
Comment (dacharyc) [inline OfflineLoginRealmAccess/OfflineLoginRealmAccess/ContentView.swift]: If I go to add more than one book, the second time the alert pops up, it still has the title from the first book. We probably need to clear the `$text` value in `addNewBook`.

```suggestion
        $books.append(newBook)
        $text.wrappedValue = ""
    }
```

File: OfflineLoginRealmAccess/OfflineLoginRealmAccess/ContentView.swift [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+        ZStack {
+            VStack {
+                Spacer()
+                VStack {
+                    TextField("Username/Email", text: $username)
+                        .disableAutocorrection(true)
+                        .textInputAutocapitalization(.none)
+                    SecureField("Password", text: $password)
+                        .textInputAutocapitalization(.never)
+                    Button("Log In") {
```
Comment (dacharyc) [inline OfflineLoginRealmAccess/OfflineLoginRealmAccess/ContentView.swift]: This is just a nitpick, but we might want to add a little padding between these buttons. They're quite close together and with just the text it was a little difficult to parse that these were separate buttons. `.padding(5)` for each button helped a lot.

File: OfflineLoginRealmAccess/OfflineLoginRealmAccess/ContentView.swift [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+                        Task {
+                            let credentials = Credentials.emailPassword(email: username, password: password)
+                            await appAuthManager.login(credentials: credentials)
+                        }
+                    }
+                    .disabled(username.isEmpty || password.isEmpty)
+                    Button("Log In Anonymously") {
```
Comment (dacharyc) [inline OfflineLoginRealmAccess/OfflineLoginRealmAccess/ContentView.swift]: It wasn't obvious to me why we had a "Log In Anonymously" button here. Is it so we can easily see "Public" books that somebody else has made, or for another reason? I wonder if we want to add a comment here about why this is in the app?

File: OfflineLoginRealmAccess/OfflineLoginRealmAccess/Managers/AppManager.swift [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+    @Published var errorMessage: String = ""
+    @Published var signUpMessage: String = ""
+
+    init() {
+        // We set a timeout for the app, so if there is no internet connection, opening the realm
+        // can throw.
+        realmApp.syncManager.timeoutOptions = SyncTimeoutOptions(connectTimeout: 5)
+    }
+
+    // Function to register a user with email and password, this will return if the registration
```
Comment (dacharyc) [inline OfflineLoginRealmAccess/OfflineLoginRealmAccess/Managers/AppManager.swift:36]:
```suggestion
    // Function to register a user with email and password. This will return if the registration
```

File: OfflineLoginRealmAccess/OfflineLoginRealmAccess/Managers/AppManager.swift [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+            try await realmApp.emailPasswordAuth.registerUser(email: username, password: password)
+            signUpMessage = "Registration successful"
+        } catch {
+            errorMessage = error.localizedDescription
+        }
+
+        isLoading = false
+    }
+
+    // Function to login a user with email and password, this will return the user if the login
```
Comment (dacharyc) [inline OfflineLoginRealmAccess/OfflineLoginRealmAccess/Managers/AppManager.swift]:
```suggestion
    // Function to login a user with email and password. This will return the user if the login
```

File: OfflineLoginRealmAccess/OfflineLoginRealmAccess/Managers/AppManager.swift [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+    // is successful or will update the error message on the UI in case of an error otherwise.
+    // Access tokens are created once a user logs in. These tokens are refreshed
+    // automatically by the SDK when needed. Manually refreshing the token is only
+    // required if requests are sent outside of the SDK. If that's the case, see:
+    // https://www.mongodb.com/docs/realm/sdk/swift/users/authenticate-users/#get-a-user-access-token
+
+    // By default, refresh tokens expire 60 days after they are issued. You can configure this
+    // time for your App's refresh tokens to be anywhere between 30 minutes and 180 days. See:
```
Comment (dacharyc) [inline OfflineLoginRealmAccess/OfflineLoginRealmAccess/Managers/AppManager.swift:62]: Should we add a note here about why refresh tokens matter? Maybe something like:

> When the refresh token expires, the user must login again.

File: OfflineLoginRealmAccess/OfflineLoginRealmAccess/Managers/RealmManager.swift [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+    func openRealmWithUser(_ user: User) async throws {
+        // Setting the configuration with `cancelAsyncOpenOnNonFatalErrors` and an initial subscription.
+        let configuration = user.flexibleSyncConfiguration(cancelAsyncOpenOnNonFatalErrors: true,
+                                                           initialSubscriptions: { subs in
+            // Setting the subscription to all public Books and the book associated to the userId
+            subs.append(QuerySubscription<Book> { $0.isPublic || $0.userId == user.id })
+        })
+
+        // Setting the download configuration to `.once` will wait to sync data only the first time the realm
+        // is opened, otherwise this will open the realm immediately and will sync any data on the background
```
Comment (dacharyc) [inline OfflineLoginRealmAccess/OfflineLoginRealmAccess/Managers/RealmManager.swift]: This is a weird English thing - we might say "in the background" or "on a background thread" but not "on the background."

```suggestion
        // is opened, otherwise this will open the realm immediately and will sync any data in the background
```

File: OfflineLoginRealmAccess/OfflineLoginRealmAccess/Managers/RealmManager.swift [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+        // if connected.
+        do {
+            // In case you want to sync your Device Sync data on every realm open, you can change this
+            // to `.always`, which in case of no internet connection will throw into opening the realm
+            // offline,
```
Comment (dacharyc) [inline OfflineLoginRealmAccess/OfflineLoginRealmAccess/Managers/RealmManager.swift]: We can probably simplify the wording of this a little bit.
```suggestion
            // If you want to sync Device Sync data on every realm open, change this
            // to `.always`. If there is no internet connection, the `catch` block tries to
            // open the realm without syncing first.
```

File: OfflineLoginRealmAccess/OfflineLoginRealmAccess/Managers/RealmManager.swift [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+            let realm = try await Realm(configuration: configuration, downloadBeforeOpen: .once)
+        } catch {
+            // In case there is an issue on the first sync, we will try to open the realm without
+            // syncing any data even on the first open.
+            // This will throw in case of the error been related to the configuration.
```
Comment (dacharyc) [inline OfflineLoginRealmAccess/OfflineLoginRealmAccess/Managers/RealmManager.swift]:
```suggestion
            // If there is an issue on the first sync, we will try to open the realm without
            // syncing any data.
            // This will throw if the error is related to the configuration.
```

**Review by dacharyc — CHANGES_REQUESTED** [review]
Unfortunately, as the app is currently written, I couldn't get past the `ProgressView()` spinner. I've suggested a handful of changes, including my fix for the ProgressView issue. But I think the example itself may not represent a real use case. Curious for your thoughts.

File: OfflineLoginRealmAccess/OfflineLoginRealmAccess/ContentView.swift (hunk 8 above)
Comment (dianaafanador3) [reply OfflineLoginRealmAccess/OfflineLoginRealmAccess/ContentView.swift]: Wanted to showcase another way to login?, also aligns with the concept of public books which is added on the logic of the example.

File: OfflineLoginRealmAccess/OfflineLoginRealmAccess/Managers/RealmManager.swift (hunk 4 above)
Comment (dianaafanador3) [reply OfflineLoginRealmAccess/OfflineLoginRealmAccess/Managers/RealmManager.swift:27]: You are right, I'm going for another approach. And if there is a configuration issue, the error will be printed

**Comment by dianaafanador3** [comment]
I changed the most important part of the code. There was actually a bug in the code which was not allowing to enter the app while offline. @dacharyc

File: OfflineLoginRealmAccess/OfflineLoginRealmAccess/Managers/AppManager.swift [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+import Foundation
+import RealmSwift
+
+let applicationServiceId = "flexiblesyncapp-mgbxq"
```
Comment (dacharyc) [inline OfflineLoginRealmAccess/OfflineLoginRealmAccess/Managers/AppManager.swift]: Looks like we need to change this back to ATLAS_APP_ID to match the README.
```suggestion
let applicationServiceId = "<ATLAS_APP_ID>"
```

File: OfflineLoginRealmAccess/OfflineLoginRealmAccess/ContentView.swift [hunk 16]
```diff
@@ -0,0 +1,215 @@
...
+
+import SwiftUI
+import RealmSwift
+
+struct ContentView: View {
+    @ObservedObject var realmManager = RealmManager()
+    @ObservedObject var app: RealmSwift.App
+
+    var body: some View {
+        // This observes changes on the App,and will trigger if the user state changes
```
Comment (dacharyc) [inline OfflineLoginRealmAccess/OfflineLoginRealmAccess/ContentView.swift]:
```suggestion
        // This observes changes on the App, and will trigger if the user state changes
```

File: OfflineLoginRealmAccess/OfflineLoginRealmAccess/ContentView.swift [hunk 17]
```diff
@@ -0,0 +1,215 @@
...
+                if let realm = realmManager.realm {
+                    BookLibraryView(user: user)
+                        .environment(\.realmConfiguration, realm.configuration)
+                } else {
+                    ProgressView()
+                        .onAppear {
+                            Task {
+                                do {
+                                    // Opens the realm asynchronously, only syncing the first time.
+                                    // In case of been offline, this will open the realm without
```
Comment (dacharyc) [inline OfflineLoginRealmAccess/OfflineLoginRealmAccess/ContentView.swift]:
```suggestion
                                    // If the device is offline, this will open the realm without
```

File: OfflineLoginRealmAccess/OfflineLoginRealmAccess/Managers/AppManager.swift [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+    // is successful or will update the error message on the UI in case of an error otherwise.
+    // Access tokens are created once a user logs in. These tokens are refreshed
+    // automatically by the SDK when needed. Manually refreshing the token is only
+    // required if requests are sent outside of the SDK. If that's the case, see:
+    // https://www.mongodb.com/docs/realm/sdk/swift/users/authenticate-users/#get-a-user-access-token
+
+    // By default, refresh tokens expire 60 days after they are issued. You can configure this
+    // time for your App's refresh tokens to be anywhere between 30 minutes and 180 days. See:
+    // https://www.mongodb.com/docs/atlas/app-services/users/sessions/#configure-refresh-token-expiration
+    // When the refresh token expires, the user must login again, and an error is throw in the
```
Comment (dacharyc) [inline OfflineLoginRealmAccess/OfflineLoginRealmAccess/Managers/AppManager.swift]:
```suggestion
    // When the refresh token expires, the user must login again, and an error is thrown in the
```

File: OfflineLoginRealmAccess/OfflineLoginRealmAccess/Managers/RealmManager.swift [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+            // If there is no internet connection, we continue to try to open the realm without syncing.
+        } catch {}
+
+        do {
+            // If there is an issue on the first sync, we will try to open the realm without
+            // syncing any data even on the first open.
+            // This will throw if the error is related to the configuration.
+            self.realm = try await Realm(configuration: configuration, downloadBeforeOpen: .never)
+        } catch {
+            // If there is other error which doesn't allow us to open the realm even without syncing,
```
Comment (dacharyc) [inline OfflineLoginRealmAccess/OfflineLoginRealmAccess/Managers/RealmManager.swift]:
```suggestion
            // If there is a different error which doesn't allow us to open the realm even without syncing,
```

**Review by dacharyc — APPROVED** [review]
There are a few small typos and we should change the app ID back to the placeholder value before merging, but otherwise this looks good!
